### ISSUE TITLE
Sub container?!

### ISSUE THREAD

**Issue by cajoy** [strongloop/loopback-component-storage#31]
I would like to split application to couple of modules so containers should have subcontainer.
Is it possible to do?

.../container1/userId/....
../container2/bookId/...

I don't feel like pushing everything in one folder is good solution. For example if I will have 100 000 containers?!

Thanks

**Comment by edlau** [comment]
Is there any update on this feature request?

**Comment by nvcken** [comment]
@raymondfeng 
I see Amazon S3 (bucket == container)
But , I use Provider google, so loopback fallback to pkgcloud handle, so container is also bucket of google cloud storage.
Then question is how can I upload to this path

container(bucket)/ folder / filename

instead of container(bucket)/ filename

Thanks

**Comment by zanemcca** [comment]
+1 For @nvcken's question

**Comment by nvcken** [comment]
I worked around by this trick.
Use "Folder_name/filename" as a Filename
Here https://github.com/strongloop/loopback-component-storage/issues/63

**Comment by zanemcca** [comment]
Thanks dude :)

**Comment by milg0** [comment]
The trick doesn't work for me, I get an error: 
```
FileSystemProvider: Invalid name:  foldername/filename.jpg
```
My configure-storage.js:
```
module.exports = function(app) {
    app.dataSources.storage.connector.getFilename = function(file, req, res) {
    	var newFileName = 'foldername/filename.jpg';  
        return newFileName;
    };
}
```
How did you manage that?

**Comment by nvcken** [comment]
That trick don't apply for FileSystem provider, it is used for Cloud Storage provider.
With FileSystem, you need manual
createContainer at beforeRemote 'upload' ref https://apidocs.strongloop.com/loopback-component-storage/#storageservice-prototype-createcontainer
Then set container name to ctx.args.req.params.container = your_container_name

**Comment by milg0** [comment]
Hm, okay. But I get the same error. :/
```
[Error: FileSystemProvider: Invalid name: test/test]
```
container.js:
```
module.exports = function(Container) {
	Container.beforeRemote('upload', function(ctx, modelInstance, next) {
		var con = Container.createContainer({
			name : "test/test"
		}, function(err, c) {
			if (err) {
				console.error(err);
			}
			else {
				console.log('Created new container', c.name);
			}
		});
		next();
	});
};
```

**Comment by nvcken** [comment]
You can try this
```
  Upload.generateContainer = function(ctx, callback) { 
    Upload.createContainer(containerOptions, function(err, container){
      ...
      callback here
    });
  }

  Upload.beforeRemote('upload', function(ctx, file, next){
    Upload.generateContainer(ctx, function(err, generatedContainer) {
      if(err) {
        next(err);
      } else {
        ctx.args.req.params.container = generatedContainer.name;
        next();
      }
    });
  });
```

**Comment by milg0** [comment]
I solved my problem with an afterRemote-Hook and moved the files with fs. I hope it helps someone out there. Be sure to use an existing new path or create it.
```
Container.afterRemote('upload', function(ctx, modelInstance, next) {
	var fs = require('fs');
	var files = ctx.result.result.files.file;

	for (var i = 0; i < files.length; i++) {
		var fileName = files[i].name;		
		
		var oldPath = './uploads/' + files[i].container + '/' + files[i].name;
		var newPath = './uploads/new/path/' + files[i].name;

		fs.readFile(oldPath, function(err, data) {
			fs.writeFile(newPath, data, function(err) {
				fs.unlink(oldPath, function() {
					if (err) {
						console.error("error unlinking", err);
						throw err;
					}
					console.log("File uploaded to: " + newPath);
				});
			});
		});
	} // for loop
	next();
});
```

**Comment by brianoconner777** [comment]
Can anyone still explain how the folder or directories are maintained using Amazon S3 in loopback.  Because I am still not able to do that.

**Comment by devpascoe** [comment]
so after a few years of dev we still can't create sub folders on AWS S3?

**Comment by nicolaspernoud** [comment]
I would really be helpful to be able to handle nested container on local filesystem, so that we could create file explorers, etc...

**Comment by moshemal** [comment]
+1

**Comment by jackrvaughan** [comment]
+1

**Comment by MiguelGonzalezAravena** [comment]
I did a fork of this project for you can create Containers with slash, that means that you can create Containers like this/isMy/newContainer, and the plugin do all the work for you. You have to replace slash with ´%2F´. All information is in https://github.com/MiguelGonzalezAravena/loopback-component-storage.
The PR is https://github.com/strongloop/loopback-component-storage/pull/217

**Comment by deepakmallah** [comment]
I have put Staack overflow question click here to [check the update](https://stackoverflow.com/questions/46756630/loopback-component-storage-upload-file-to-sub-directory-of-container).

**Comment by sminodonte** [comment]
+1

**Comment by rendomnet** [comment]
is this fixed?

**Comment by regevbr** [comment]
I have created a different component where the :container is actually the root folder in the bucket and the bucket is defined in the data source level. Also this component supports acl and relations which makes more sense - [loopback-component-flatstorage](https://github.com/regevbr/loopback-component-flatstorage)
Regarding adding nested folders I guess that using %2F in the file name will work but I didnt try it

**Comment by kjdelisle** [comment]
@raymondfeng had some concerns regarding the last pull request and its implementation. Though @ericdowell has said last week that he's interested in giving it a go, I would still encourage others to take a shot at it (or coordinate with one another).

TL;DR: pull requests welcome! :D

**Comment by ericdowell** [comment]
@kjdelisle After consider a few different solutions, I've found that there is only one clean way to solve this issue. To achieve this clean solution we would need to partially break several of the remote method function parameters (aka not backward compatible). Note that all HTTP remote api calls would continue to work the same (aka calls to `GET /api/container/`).

I can add this feature and would really like to since we use this component at work. But before I continue further work I'd like to know more about what kinds of backward compatibility I need to worry about. The main issue is that we need to add more parameters to the remote method functions. The new parameters would either pull `prefix` query string or we'd pass the whole `req` object and check for the `prefix` query string that way. I would pick `prefix` over say `folder` because of the popularity of AWS S3. This `prefix` parameter would be optional and everything would continue to work without it.

The other part of the solution is to allow for users to implement their specific provider. Similar to how there is a way to override the filename with `getFilename`, I would add a `setPrefix` or `setFolderName` function. This function would take in similar parameters as `getFilename`.

TL;DR:
@kjdelisle I would like to start a discussion on what kind(s) of backward compatibility breaks will be allowed to implement this feature.

**Comment by louiepiol** [comment]
+1

**Comment by jackrvaughan** [comment]
Commenting as I don't want this to go stale. 

This hasn't been resolved yet, correct? If so, what is the recommended way of accomplishing this? I see a few different potential solutions above.

**Comment by demiro** [comment]
bump

**Comment by regevbr** [comment]
@demiro see my comment above - "I have created a different component where the :container is actually the root folder in the bucket and the bucket is defined in the data source level. Also this component supports acl and relations which makes more sense - loopback-component-flatstorage (https://github.com/regevbr/loopback-component-flatstorage) 
Regarding adding nested folders I guess that using %2F in the file name will work but I didnt try it"

**Comment by badgujargaurav** [comment]
@regevbr I'm quite new in loopback community and I encountered with this same issue. Also, I saw your git repo but due to lack of experience don't know how to use your code. Can you please help me in solving it?

**Comment by jackrvaughan** [comment]
@badgujargaurav - This may be helpful: 

https://medium.com/@jackrobertscott/upload-files-to-aws-s3-in-loopback-29a3f01119f4

**Comment by regevbr** [comment]
Sorry on vacation will address you as soon as i can

**Comment by RAlhawash** [comment]
any update on this ?

**Comment by pookdeveloper** [comment]
Same error I want a subfolder. it's basic for structure data, please add it, thanks

**Comment by regevbr** [comment]
@pookdeveloper you can use my lib for now - https://github.com/regevbr/loopback-component-flatstorage

**Comment by pookdeveloper** [comment]
Thanks @regevbr
